### PR TITLE
refactor: move auth, session, and relay logic into providers

### PR DESCRIPTION
**Pull request by sarahzdev**
**Summary**
- Refactor authentication logic for Google, Discord, EthWallet (now EthereumAccount) by moving it outside of `LitAuthClient` and into their respective provider classes
- Code related to handling authentication, session sigs, and relay server has been reorganized into the corresponding provider classes
- Added WebAuthn support

Notes
- Added `moduleResolution: "node16"` to handle dynamically import `@simplewebauthn/browser`, which is ESM-only (more context on the issue/workaround [here](https://github.com/microsoft/TypeScript/issues/43329)
- This flag breaks the spec so commented the spec for now
- Porting README content to the dev docs

TODO:
[x] Update spec test
[x] Update demo app
[x] Update readme
[ ] Merge into PR #67

### CODE REVIEW

File: packages/lit-auth-client/src/lib/lit-auth-client.ts [hunk 1]
```diff
@@ -82,385 +65,102 @@ export class LitAuthClient {
...
-      signedMessage: toSign,
-      address: params.address,
-    };
-
-    const authMethod = {
-      authMethodType: AuthMethodType.EthWallet,
-      accessToken: JSON.stringify(authSig),
-    };
-    return authMethod;
+  initDiscordProvider(params: InitOAuthProviderOptions): DiscordProvider {
```
Comment (joshLong145) [inline packages/lit-auth-client/src/lib/lit-auth-client.ts:106]: Perhaps this shouldn't be explicitly defined for each provider type.
maybe instead there is a single `initProvider<T extends ProviderOptions>(params: ProviderOptions<T>)`
T in the `ProviderOptions` will be a concrete interface type for the Provider Options. A bit cleaner than having an explicit init.

File: packages/lit-auth-client/src/lib/relay.ts [hunk 2]
```diff
@@ -162,6 +162,34 @@ export class LitRelay implements IRelay {
...
   }
 
+  /**
+   * Generate options for registering a new credential to pass to the authenticator
+   *
+   * @param {string} [username] - Optional username to associate with the credential
+   *
+   * @returns {Promise<any>} Registration options for the browser to pass to the authenticator
+   */
+  public async generateRegistrationOptions(username?: string): Promise<any> {
```
Comment (joshLong145) [inline packages/lit-auth-client/src/lib/relay.ts:172]: Should this be defined on `IRelay`

File: packages/lit-auth-client/src/lib/providers/WebAuthnProvider.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      rpId,
+    };
+
+    // Authenticate with WebAuthn
+    const authenticationResponse = await startAuthentication(
+      authenticationOptions
+    );
+
+    // BUG: We need to make sure userHandle is base64url encoded. Deep copy the authentication response.
+    const actualAuthenticationResponse = JSON.parse(
```
Comment (joshLong145) [inline packages/lit-auth-client/src/lib/providers/WebAuthnProvider.ts:106]: Is this a bug, the spec does require the `userId` to be base64 encoded.

File: packages/lit-auth-client/src/lib/providers/WebAuthnProvider.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  /**
+   * Authenticate with WebAuthn
+   *
+   * @param {string} [rpcUrl] - RPC URL to use for getting latest block hash
+   *
+   * @returns {Promise<AuthMethod>} - Auth method object containing WebAuthn auth data
+   */
+  public async authenticate(): Promise<AuthMethod> {
```
Comment (joshLong145) [inline packages/lit-auth-client/src/lib/providers/WebAuthnProvider.ts:81]: should we add this method signature to `baseProvider`? might be useful for other auth methods in the future.

File: packages/lit-auth-client/src/lib/providers/DiscordProvider.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    // Redirect to login url
+    window.location.assign(loginUrl);
+  }
+
+  /**
+   * Validate the URL parameters returned from Lit's login server and return the OAuth token
+   *
+   * @returns {AuthMethod} - Auth method object containing the OAuth token
+   */
+  public authenticateWithLit(): AuthMethod {
```
Comment (joshLong145) [inline packages/lit-auth-client/src/lib/providers/DiscordProvider.ts:39]: perhaps we uniform this to `authenticate` and add to `BaseProvider`

**Review by joshLong145 — CHANGES_REQUESTED** [review]

File: packages/lit-auth-client/src/lib/lit-auth-client.ts (hunk 1 above)
Comment (sarahzdev) [reply packages/lit-auth-client/src/lib/lit-auth-client.ts]: @joshLong145 - updated, thanks!

File: packages/lit-auth-client/src/lib/providers/WebAuthnProvider.ts (hunk 3 above)
Comment (sarahzdev) [reply packages/lit-auth-client/src/lib/providers/WebAuthnProvider.ts:101]: @joshLong145  - updated the comment and `userHandle` check slightly. looks like the `@simplewebauthn/browser` pkg we’re using returns the `userhandle` as a regular string seen [here](https://github.com/MasterKale/SimpleWebAuthn/blob/76c7a46b60b1093e9e71f38670175bee45512473/packages/typescript-types/src/index.ts#L145)

File: packages/lit-auth-client/src/lib/providers/WebAuthnProvider.ts (hunk 4 above)
Comment (sarahzdev) [reply packages/lit-auth-client/src/lib/providers/WebAuthnProvider.ts:77]: @joshLong145 - added an abstract method `authenticate` to `BaseProvider`, ty!

File: packages/lit-auth-client/src/lib/providers/DiscordProvider.ts (hunk 5 above)
Comment (sarahzdev) [reply packages/lit-auth-client/src/lib/providers/DiscordProvider.ts]: @joshLong145 - updated this to implement the `BaseProvider`'s `authenticate`

File: packages/lit-auth-client/src/lib/providers/WebAuthnProvider.ts (hunk 3 above)
Comment (joshLong145) [reply packages/lit-auth-client/src/lib/providers/WebAuthnProvider.ts:101]: Oh I see, that's good to know.

**Review by joshLong145 — APPROVED** [review]

**Pull request merged**